src/preact/use-is-intersecting.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../log';
+import {useEffect, useState} from './index';
+
+/** @type {null|IntersectionObserver} */ let sharedObserver = null;
+
+/**
+ * @param {{current: HTMLElement}} ref
+ * @return {boolean}
+ */
+export function useIsIntersecting(ref) {
+  const {0: isIntersecting, 1: setIsIntersecting} = useState(null);
+  if (sharedObserver === null) {
+    sharedObserver = new IntersectionObserver(entries => {
+      const last = entries[entries.length - 1];
```

Comment:
Actually, this could have been caused by any one of the elements coming into viewport, since it's shared. Not to mention, the `set` is tied to a single element. We need to store these in a map, and call the correct set based on the element that's triggering the entries (and they can be entries for multiple elements).

Reply:
It could even be a `Set` right? Something like: 

```js
setIntersections(new Set(entries.map(e => e.target))); // Makes a Set<!Element>
```

Then later the return statement would look like:
```js
return intersections.has(ref.current)
```

Reply:
@samouri Is `setIntersections` intended to be defined at the doc-level? I am getting obscure type errors with this approach for now, so let's discuss offline to make sure I'm understanding you correctly. 😅